typescript/ts-lowerings/src/org/jetbrains/dukat/tsLowerings/IntroduceMissingConstructors.kt

```diff
@@ -1,45 +1,74 @@
 import org.jetbrains.dukat.ownerContext.NodeOwner
 import org.jetbrains.dukat.toposort.toposort
+import org.jetbrains.dukat.tsLowerings.ResolveTypeParamsInConstructor
 import org.jetbrains.dukat.tsLowerings.TopLevelDeclarationResolver
 import org.jetbrains.dukat.tsmodel.ClassDeclaration
 import org.jetbrains.dukat.tsmodel.ConstructorDeclaration
 import org.jetbrains.dukat.tsmodel.ModuleDeclaration
 import org.jetbrains.dukat.tsmodel.SourceSetDeclaration
 import org.jetbrains.dukat.tsmodel.TopLevelDeclaration
+import org.jetbrains.dukat.tsmodel.types.ParameterValueDeclaration
 
 class IntroduceMissingConstructors : TopLevelDeclarationLowering {
     private lateinit var declarationResolver: TopLevelDeclarationResolver
-    private val parentConstructorsMap = mutableMapOf<String, List<ConstructorDeclaration>>()
-
-    override fun lowerClassDeclaration(declaration: ClassDeclaration, owner: NodeOwner<ModuleDeclaration>?): TopLevelDeclaration? {
-        val constructors = parentConstructorsMap[declaration.uid]
-        val declarationResolved = if (constructors.isNullOrEmpty() || declaration.parentEntities.isEmpty()) {
-            declaration
-        } else {
-            // TODO: this is a strange thing to do and I need to figure out how to avoid this
-            declaration.copy(members = (constructors + declaration.members).distinct())
+    private val parentConstructorsMap = mutableMapOf<String, ClassWithConstructors>()
+
+    private data class ClassWithConstructors(
+        val classDeclaration: ClassDeclaration,
+        val constructors: List<ConstructorDeclaration>
+    )
+
+    private fun ClassWithConstructors?.addConstructorsInto(declaration: ClassDeclaration): ClassDeclaration {
+        if (this == null || constructors.isEmpty() || declaration.parentEntities.isEmpty()) {
+            return declaration
+        }
+        // TODO: this is a strange thing to do and I need to figure out how to avoid this
+        return declaration.copy(members = (constructors + declaration.members).distinct())
+    }
+
+    private fun ClassWithConstructors.getConstructorsWithResolvedTypeParams(actualTypes: List<ParameterValueDeclaration>) =
+        when {
+            classDeclaration.typeParameters.isEmpty() -> constructors
+            else -> {
+                val typesMapping = classDeclaration.typeParameters.withIndex().map { (index, type) ->
+                    val resolvedType = actualTypes.getOrNull(index) ?: type.defaultValue
+                    type.name to resolvedType!!
```

Comment:
How should we handle this kind of situation?
Ideally, the typescript compiler should show a compilation error for invalid declarations, but, seems like, we try to handle even invalid type definitions.
P.S. In the example we try to resolve type parameter with provided type or with default value for the type, but, we can be faced with the invalid type declaration which doesn't contain a default type value and an actual type was not provided.